File: client/store/modules/subject.js
const state = () => ({
  subjects: [],
})

const getters = {
  subjectById: state => subjectId => {
    return state.subjects.find(s => s._id === subjectId)
  },

  subjectName: state => subjectId => {
    const subject = state.subjects.find(s => s._id === subjectId)
    if (subject) {
      if (subject.info && subject.info["Имя"])
        return subject.info["Имя"]
      else
        return subject._id
    }
    return ''
  },
}

const mutations = {
  SET_SUBJECTS(state, subjects) {
    state.subjects = subjects
  },

  REMOVE_SUBJECT(state, subjectId) {
    state.subjects = state.subjects.filter(r => r._id !== subjectId)
  },

  ADD_SUBJECT(state, subject) {
    state.subjects.push(subject)
  }
}

const actions = {
  async createSubject({ commit }, { researchId, ...subject }) {
    const res = await this.$axios.$post(`/research/${researchId}/subject`, {
      ...subject
    })
    commit('ADD_SUBJECT', res)

    return res
  },

  async getSubjects({ commit }, { researchId }) {
    const res = await this.$axios.$get(`/research/${researchId}/subject`)
    commit('SET_SUBJECTS', res)

    return res
  },

  async deleteSubject({ commit }, { subjectId }) {
    const res = await this.$axios.$delete(`/subject/${subjectId}`)
    commit('REMOVE_SUBJECT', subjectId)

    return res
  }
}

export default {
  state,
  getters,
  mutations,
  actions,
}

File: server/api/subject.py
from flask import Blueprint, request, jsonify
from service import subject_service

import utils.flie_system as fs
import traceback


def get_subject_api():
    router = Blueprint('subject_router', __name__)

    @router.route("/research/<research_id>/subject", methods=['GET'])
    def get_research_subjects(research_id):
        res = [
            dict(**subject, directory=subject_service.get_subject_dir(research_id, subject['_id']))
            for subject in subject_service.get_subjects(research_id)
        ]
        return jsonify(res)

    @router.route("/research/<research_id>/subject", methods=['POST'])
    def create_subject(research_id):
        subject_data = request.json
        subject = subject_service.create_subject(research_id, subject_data)
        if subject:
            fs.create_directory(subject_service.get_subject_dir(research_id, subject['_id']))
        return jsonify(subject)

    @router.route('/subject/<subject_id>', methods=['DELETE'])
    def delete_subject(subject_id):
        subject = subject_service.get_subject(subject_id)
        if subject:
            subject_dir = subject_service.get_subject_dir(subject['research_id'], subject_id)
            fs.delete_directory(subject_dir)
            subject_service.delete_subject(subject_id)
            return jsonify(True)
        return jsonify(False)

    return router
